File: src/main/java/com/beyondbell/bugisoft/old/Standalone/Commands/Ping/Ping.java
package com.beyondbell.bugisoft.old.Standalone.Commands.Ping;

import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.entities.Message;
import net.dv8tion.jda.core.entities.MessageEmbed;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;

public final class Ping {
	private static volatile long startingTime = System.currentTimeMillis();
	private static volatile boolean released = true;
	private static Message message;

	public static void ping(final GuildMessageReceivedEvent event) {
		if (released) {
			released = false;
			message = event.getChannel().sendMessage("Ping Calculating...").complete();
			startingTime = System.currentTimeMillis();
		} else {
			event.getChannel().sendMessage("A Ping Check is Already in Process").queue();
		}
	}

	public static void pingReceived(final GuildMessageReceivedEvent event) {
		final MessageEmbed messageEmbed = new EmbedBuilder()
				.addField("Bot to Server Ping", String.valueOf(System.currentTimeMillis() - startingTime), true)
				.addField("Gateway Ping", String.valueOf(event.getJDA().getPing()), true)
				.addField("Server Ping", String.valueOf(System.currentTimeMillis() - startingTime - event.getJDA().getPing()), true)
				.build();
		event.getChannel().sendMessage(messageEmbed).queue();
		event.getChannel().deleteMessageById(message.getId()).queue();

		message = null;
		released = true;
	}
}

File: src/main/java/com/beyondbell/pingDiscordBot/Ping.kt
package com.beyondbell.pingDiscordBot

import net.dv8tion.jda.core.EmbedBuilder
import net.dv8tion.jda.core.JDA
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.core.hooks.ListenerAdapter
import java.util.concurrent.TimeUnit

object Ping : ListenerAdapter() {
	internal val times: Array<Long?> = arrayOf(null, null, null, null, null)
	private var outboundRequests = 0

	override fun onGuildMessageReceived(event: GuildMessageReceivedEvent) {
		if (event.author.isBot && event.message.contentRaw.startsWith("Ping Request: ")) {
			val totalPing = System.currentTimeMillis() - times[Integer.valueOf(event.message.contentRaw.removePrefix("Ping Request: "))] as Long

			event.message.editMessage(EmbedBuilder()
					.setTitle("Ping Request Receipt")
					.addField("Current Total Ping", totalPing.toString(), true)
					.addField("First Gateway Ping", event.jda.ping.toString(), true)
					.addField("Estimated Server Ping", (totalPing - event.jda.ping).toString(), true)
					.build()).complete()
			event.message.delete().queueAfter(1, TimeUnit.MINUTES)

			times[event.message.contentRaw.removePrefix("Ping Request: ").toInt()] = null

			changeListenerStatus(event.jda, false)
		}
	}

	internal fun getNextId(): Int {
		for (i in 0 until times.size) {
			if (times[i] == null) {
				return i
			}
		}
		return -1
	}

	internal fun changeListenerStatus(jda: JDA, register: Boolean) {
		if (register) {
			if (outboundRequests++ == 0) {
				jda.addEventListener(this)
			}
		} else {
			if (--outboundRequests == 0) {
				jda.removeEventListener(this)
			}
		}
	}
}